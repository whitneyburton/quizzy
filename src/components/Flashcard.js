import React, { Component } from 'react';
import '../styles/main.scss';

export default class Flashcard extends Component {
  constructor() {
    super();
    this.state = {
      correct: null
    }
  }

  validateAnswer = (e) => {
    let { flashcard, saveToStorage, removeCorrectFromStorage } = this.props;
    let answerClicked = e.target.innerText;
    if (flashcard.answer === answerClicked) {
      this.setState({ correct: true });
      removeCorrectFromStorage(flashcard);
    } else {
      this.setState({ correct: false });
      saveToStorage(flashcard);
    }
  }

  randomizeAnswers() {
    let { flashcard, flashcards } = this.props;
    let allAnswers = flashcards
      .map(flashcard => flashcard.answer)
      .sort(() => 0.5 - Math.random())
    let correctAnswerIndex = allAnswers.indexOf(flashcard.answer)
    allAnswers.splice(correctAnswerIndex, 1)
    let finalArray = [...allAnswers
      .splice(0, 2), flashcard.answer]
      .sort(() => 0.5 - Math.random())
    return finalArray;
  }

  render() {
    const { flashcard, filteredCards } = this.props;
    const { correct } = this.state;
    let currentAnswersArray = this.randomizeAnswers();
    let flashcardClass = 'Flashcard';
    let correctFeedback;
    let incorrectFeedback;
    if (correct === true) {
      flashcardClass = 'Flashcard correct-answer'
      correctFeedback = 'show'
      incorrectFeedback = 'hide'
    } else if (correct === false) {
      flashcardClass = 'Flashcard incorrect-answer'
      incorrectFeedback = 'show'
      correctFeedback = 'hide'
    } else {
      incorrectFeedback = 'hide'
      correctFeedback = 'hide'
    }

    return (
      <div className={flashcardClass}>
        <p className="question-counter">
          Question {filteredCards.indexOf(flashcard) + 1}/{filteredCards.length}
        </p>
        <p className="flashcard-question">{flashcard.question}</p>
        <div className="buttons-container">
          {currentAnswersArray.map((answer, index) => {
            return <button
              key={index}
              onClick={this.validateAnswer}
              className="buttons flashcard-buttons"
              type="button">{answer}</button>
          })}
        </div>
        <p className={correctFeedback}>That's right! Nice work, Quizzy pro.</p>
        <p className={incorrectFeedback}>Not quite. Keep guessing!</p>
        <div className="learn-more">
          <a
            href={flashcard.mdn_link}
            className="mdn-link"
            target="_blank"
            rel="noopener noreferrer">Learn More</a>
        </div>
      </div>
    )
  }
}