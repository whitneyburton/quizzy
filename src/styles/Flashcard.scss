.Flashcard {
  border: 1px solid black;
  border-radius: 5px;
  background-color: #ccdedc;
  width: 75%;
  // height: 5000px;
  margin: 10px 0px;
  padding: 28px;
}

.question-counter {
  text-align: right;
  font-size: 12px;
}

.flashcard-question {
  margin: 20px 0px;
}

.buttons-container {
  display: flex;
  justify-content: space-evenly;
}

.flashcard-buttons {
  padding: 10px;
}

.hide {
  display: none;
}