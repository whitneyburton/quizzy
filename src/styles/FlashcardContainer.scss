.FlashcardContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: whitesmoke;
  height: 84vh;
  width: 60%;
  margin: 0px auto;
  overflow: scroll;
}

.all-flashcards-header {
  padding: 7% 7% 3% 7%;
}

@media screen and (max-width: 755px) {
  .FlashcardContainer {
    width: 90%;
    height: 60vh;
  }
}