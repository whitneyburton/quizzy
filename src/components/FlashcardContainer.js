import React from 'react';
import Flashcard from './Flashcard'
import '../styles/main.scss';

export default function FlashcardContainer() {
    return (
      <div className="FlashcardContainer">
        <h2 className="all-flashcards-header">
          Flashcards for: <span>Mutator</span> methods
          </h2>
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />
        <Flashcard />

    </div>
  );
}