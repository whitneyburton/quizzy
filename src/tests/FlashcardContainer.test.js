import React from 'react';
import FlashcardContainer from '../Components/FlashcardContainer';
import { shallow } from 'enzyme';
import Flashcard from '../components/Flashcard';

describe('FlashcardContainer', () => {
  let wrapper;
  const category = 'All Methods!'
  const filteredCards = [{}, {}, {}]
  const saveToStorageMock = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <FlashcardContainer
        category={category}
        filteredCards={filteredCards}
        saveToStorage={saveToStorageMock}
      />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })
})