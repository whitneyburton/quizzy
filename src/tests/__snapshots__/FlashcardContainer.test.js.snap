// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlashcardContainer should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FlashcardContainer
    category="All Methods!"
    filteredCards={
      Array [
        Object {},
        Object {},
        Object {},
      ]
    }
    saveToStorage={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="all-flashcards-header"
        >
          All Methods!
        </h2>,
        Array [
          <Flashcard
            filteredCards={
              Array [
                Object {},
                Object {},
                Object {},
              ]
            }
            flashcard={Object {}}
            saveToStorage={[MockFunction]}
          />,
          <Flashcard
            filteredCards={
              Array [
                Object {},
                Object {},
                Object {},
              ]
            }
            flashcard={Object {}}
            saveToStorage={[MockFunction]}
          />,
          <Flashcard
            filteredCards={
              Array [
                Object {},
                Object {},
                Object {},
              ]
            }
            flashcard={Object {}}
            saveToStorage={[MockFunction]}
          />,
        ],
      ],
      "className": "FlashcardContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "All Methods!",
          "className": "all-flashcards-header",
        },
        "ref": null,
        "rendered": "All Methods!",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "filteredCards": Array [
            Object {},
            Object {},
            Object {},
          ],
          "flashcard": Object {},
          "removeCorrectFromStorage": undefined,
          "saveToStorage": [MockFunction],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "filteredCards": Array [
            Object {},
            Object {},
            Object {},
          ],
          "flashcard": Object {},
          "removeCorrectFromStorage": undefined,
          "saveToStorage": [MockFunction],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "filteredCards": Array [
            Object {},
            Object {},
            Object {},
          ],
          "flashcard": Object {},
          "removeCorrectFromStorage": undefined,
          "saveToStorage": [MockFunction],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="all-flashcards-header"
          >
            All Methods!
          </h2>,
          Array [
            <Flashcard
              filteredCards={
                Array [
                  Object {},
                  Object {},
                  Object {},
                ]
              }
              flashcard={Object {}}
              saveToStorage={[MockFunction]}
            />,
            <Flashcard
              filteredCards={
                Array [
                  Object {},
                  Object {},
                  Object {},
                ]
              }
              flashcard={Object {}}
              saveToStorage={[MockFunction]}
            />,
            <Flashcard
              filteredCards={
                Array [
                  Object {},
                  Object {},
                  Object {},
                ]
              }
              flashcard={Object {}}
              saveToStorage={[MockFunction]}
            />,
          ],
        ],
        "className": "FlashcardContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "All Methods!",
            "className": "all-flashcards-header",
          },
          "ref": null,
          "rendered": "All Methods!",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "filteredCards": Array [
              Object {},
              Object {},
              Object {},
            ],
            "flashcard": Object {},
            "removeCorrectFromStorage": undefined,
            "saveToStorage": [MockFunction],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "filteredCards": Array [
              Object {},
              Object {},
              Object {},
            ],
            "flashcard": Object {},
            "removeCorrectFromStorage": undefined,
            "saveToStorage": [MockFunction],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "filteredCards": Array [
              Object {},
              Object {},
              Object {},
            ],
            "flashcard": Object {},
            "removeCorrectFromStorage": undefined,
            "saveToStorage": [MockFunction],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
